File: neetcode/heap/Task Scheduler.py
# https://leetcode.com/problems/task-scheduler/description/

from collections import deque,Counter
import heapq
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        count = Counter(tasks)
        # heap 
        heap =[-i for i in count.values()]
        heapq.heapify(heap)
        
        # q
        q = deque()
        time = 0
        while q or heap:
            if heap:
                value = heapq.heappop(heap) * -1
                value = value - 1
                if value > 0:
                    q.append((value, time+n))
            while q and q[0][1] <= time:
                heapq.heappush(heap,q.popleft()[0] * -1)     
            time+=1
        return time

File: neetcode/go/heap/Task Scheduler.go
package heap

func leastInterval(tasks []byte, n int) int {
	count := make(map[byte]int)
	for _, task := range tasks {
		count[task]++
	}

	maxHeap := priorityqueue.NewWith(func(a, b interface{}) int {
		return b.(int) - a.(int)
	})
	for _, cnt := range count {
		maxHeap.Enqueue(cnt)
	}

	time := 0
	q := make([][2]int, 0)

	for maxHeap.Size() > 0 || len(q) > 0 {
		time++

		if maxHeap.Size() == 0 {
			time = q[0][1]
		} else {
			cnt, _ := maxHeap.Dequeue()
			cnt = cnt.(int) - 1
			if cnt.(int) > 0 {
				q = append(q, [2]int{cnt.(int), time + n})
			}
		}

		if len(q) > 0 && q[0][1] == time {
			maxHeap.Enqueue(q[0][0])
			q = q[1:]
		}
	}

	return time
}
